File: system-test/dotnet/Dsl.Core/Scenario/When/Steps/Base/BaseWhenStep.cs
using Dsl.Core.Scenario.Then;
using Dsl.Port.Then;
using Dsl.Port.When.Steps.Base;
using Dsl.Core.Shared;
using Driver.Adapter;
using Optivem.Testing;

namespace Dsl.Core.Scenario.When.Steps.Base;

public abstract class BaseWhen<TSuccessResponse, TSuccessVerification>
    : IWhenStep
    where TSuccessVerification : ResponseVerification<TSuccessResponse>
{
    private readonly UseCaseDsl _app;
    private readonly ScenarioDsl _scenario;
    private readonly Func<Task> _ensureGiven;

    protected BaseWhen(UseCaseDsl app, ScenarioDsl scenario, Func<Task> ensureGiven)
    {
        _app = app;
        _scenario = scenario;
        _ensureGiven = ensureGiven;
    }

    public ThenStage<TSuccessResponse, TSuccessVerification> Then()
    {
        return new ThenStage<TSuccessResponse, TSuccessVerification>(Channel, _app, async () =>
        {
            await _ensureGiven();
            return await Execute(_app);
        });
    }

    IThenResultStage IWhenStep.Then() => Then();

    protected abstract Task<ExecutionResult<TSuccessResponse, TSuccessVerification>> Execute(UseCaseDsl app);

    protected Channel Channel => _scenario.Channel;
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/dsl/core/scenario/when/steps/BaseWhenStep.java
package com.mycompany.myshop.testkit.dsl.core.scenario.when.steps;

import com.mycompany.myshop.testkit.dsl.core.shared.ResponseVerification;
import com.mycompany.myshop.testkit.dsl.core.usecase.UseCaseDsl;
import com.mycompany.myshop.testkit.dsl.core.scenario.ExecutionResult;
import com.mycompany.myshop.testkit.dsl.core.scenario.then.ThenResultImpl;

public abstract class BaseWhenStep<R, V extends ResponseVerification<R>> {
    private final UseCaseDsl app;

    protected BaseWhenStep(UseCaseDsl app) {
        this.app = app;
    }
    public ThenResultImpl<R, V> then() {
        var result = execute(app);
        return new ThenResultImpl<>(app, result);
    }

    protected abstract ExecutionResult<R, V> execute(UseCaseDsl app);
}



